pkg/utils/concurrency.go

```diff
@@ -0,0 +1,110 @@
+package utils
+
+import (
+	"sync"
+)
+
+// TODO: closeWhenDone bool parameter ::
+//			It currently is experimental, and therefore exists.
+//			Is there ever a situation to use false?
+
+func SliceOfChannelsRawMerger[IR any, MR any](individualResultChannels []<-chan IR, outputChannel chan<- MR, mappingFn func(IR) (MR, error), closeWhenDone bool) *sync.WaitGroup {
+	var wg sync.WaitGroup
+	wg.Add(len(individualResultChannels))
+	for _, irc := range individualResultChannels {
+		go func(c <-chan IR) {
+			for r := range c {
+				mr, err := mappingFn(r)
+				if err == nil {
+					outputChannel <- mr
+				}
+			}
+			wg.Done()
+		}(irc)
+	}
+	if closeWhenDone {
+		go func() {
+			wg.Wait()
+			close(outputChannel)
+		}()
+	}
+
+	return &wg
+}
+
+func SliceOfChannelsRawMergerWithoutMapping[T any](individualResultsChannels []<-chan T, outputChannel chan<- T, closeWhenDone bool) *sync.WaitGroup {
+	return SliceOfChannelsRawMerger(individualResultsChannels, outputChannel, func(v T) (T, error) { return v, nil }, closeWhenDone)
+}
+
+func SliceOfChannelsMergerWithErrors[IV any, OV any](individualResultChannels []<-chan ErrorOr[IV], successChannel chan<- OV, errorChannel chan<- error, mappingFn func(IV) (OV, error), closeWhenDone bool) *sync.WaitGroup {
+	var wg sync.WaitGroup
+	wg.Add(len(individualResultChannels))
+	for _, irc := range individualResultChannels {
+		go func(c <-chan ErrorOr[IV]) {
+			for r := range c {
+				if r.Error != nil {
+					errorChannel <- r.Error
+				} else {
+					mv, err := mappingFn(r.Value)
+					if err != nil {
+						errorChannel <- err
+					} else {
+						successChannel <- mv
+					}
+				}
+			}
+			wg.Done()
+		}(irc)
+	}
+	if closeWhenDone {
+		go func() {
+			wg.Wait()
+			close(successChannel)
+			close(errorChannel)
+		}()
+	}
+	return &wg
+}
+
+func SliceOfChannelsTransformer[IV any, OV any](inputChanels []<-chan IV, mappingFn func(v IV) OV) (outputChannels []<-chan OV) {
+	rawOutputChannels := make([]<-chan OV, len(inputChanels))
+
+	for ci, c := range inputChanels {
+		roc := make(chan OV)
+		go func(i int, ic <-chan IV, oc chan OV) {
+			for iv := range ic {
+				oc <- mappingFn(iv)
+			}
+			close(oc)
+		}(ci, c, roc)
+		rawOutputChannels[ci] = roc
+	}
+
+	outputChannels = rawOutputChannels
+	return
+}
+
+func SliceOfChannelsReducer[IV any, OV any](individualResultsChannels []<-chan IV, outputChannel chan<- OV,

```

Comment:
I can't really see the benefit of using generics here that could not be simply replaced with a simple interface usage. Do we need generics at all? it makes the code quite hard to grasp here

Reply:
I'll be the first to admit that I have an affinity for generics :)

However, in the case of this concurrency bundle, I elected for generics in order to better represent their intended usage.

To me at least, a function with a signature that accepts and emits interface{} is saying to the programmer "this function accepts anything as input" and that the mapping or reducing function can turn it to "anything else"

I intended these functions to convey a similar, but more restrictive contract: while they still can be used for any arbitrary pair of input and output types, the relationship between them is fixed. All input, no matter how many channels are being combined, is of the same type, and similarly, the type required by the consumer of the returned channels is statically guaranteed to be of the correct type as well.

Personally, I'd prefer to improve the documentation and naming of the types to make this more clear versus switching to interface{}, but I would love to hear your opinions as well!

Reply:
While I'm not sure my first stab at per-fn comments are good enough yet, I also refactored the opaque looking generic types to improve the autocomplete experience a little bit.